pandas/core/arrays/numpy_.py

```diff
@@ -0,0 +1,327 @@
+        # That's probably a bad idea.
+        result = np.asarray(scalars)
+        if copy and result is scalars:
+            result = result.copy()
+        return cls(result)
+
+    @classmethod
+    def _from_factorized(cls, values, original):
+        return cls(values)
+
+    @classmethod
+    def _concat_same_type(cls, to_concat):
+        return cls(np.concatenate(to_concat))
+
+    # ------------------------------------------------------------------------
+    # Data
+
+    @property
+    def dtype(self):
+        return self._dtype
+
+    # ------------------------------------------------------------------------
+    # NumPy Array Interface
+
+    def __array__(self, dtype=None):
+        return np.asarray(self._ndarray, dtype=dtype)
+
+    # TODO: __array_ufunc__
+
+    # ------------------------------------------------------------------------
+    # Pandas ExtensionArray Interface
+
+    def __getitem__(self, item):
+        if isinstance(item, type(self)):
+            item = item._ndarray
+
+        result = self._ndarray[item]
+        if not lib.is_scalar(result):
+            result = type(self)(result)
+        return result
+
+    def __setitem__(self, key, value):
+        value = extract_array(value, extract_numpy=True)
+
+        if not lib.is_scalar(key) and is_list_like(key):
+            key = np.asarray(key)
+            if not len(key):
+                # early return to avoid casting unnecessarily.
+                return
+
+        if not lib.is_scalar(value):
+            value = np.asarray(value)
+
+        values = self._ndarray
+        t = np.result_type(value, values)
+        if t != self._ndarray.dtype:
+            values = values.astype(t, casting='safe')
+            values[key] = value
+            self._dtype = PandasDtype(t)
+            self._ndarray = values
+        else:
+            self._ndarray[key] = value
+
+    def __len__(self):
+        return len(self._ndarray)
+
+    @property
+    def nbytes(self):
+        return self._ndarray.nbytes
+
+    def isna(self):
+        from pandas import isna
+
+        return isna(self._ndarray)
+
+    def fillna(self, value=None, method=None, limit=None):
+        from pandas.api.types import is_array_like
+        from pandas.util._validators import validate_fillna_kwargs
+        from pandas.core.missing import pad_1d, backfill_1d
+
+        # TODO(_values_for_fillna): remove this
+        value, method = validate_fillna_kwargs(value, method)
+
+        mask = self.isna()
+
+        if is_array_like(value):
+            if len(value) != len(self):
+                raise ValueError("Length of 'value' does not match. Got ({}) "
+                                 " expected {}".format(len(value), len(self)))
+            value = value[mask]
+
+        if mask.any():
+            if method is not None:
+                func = pad_1d if method == 'pad' else backfill_1d
+                new_values = func(self._ndarray, limit=limit,
```

Comment:
Is this `self._ndarray` the only difference with the base fillna implementation?

Reply:
Yup. we have an issue somewhere for this.

Reply:
What's the question? `_ndarray` is the attribute the data is stored under: https://github.com/pandas-dev/pandas/pull/24227/files#diff-c42d18d11951747332a3104156c139c4R135